sugargame

ayushnawal opened the pull request:
tested

ayushnawal committed:
update sugargame to v1.3

ayushnawal committed:
add license and repository metadata

quozl commented:
Thanks, but given your other pull requests involving the Sugargame API, I'd prefer to see you fix them before I review any further changes.

ayushnawal commented:
I reviewed and fixed them as you suggested, you can review these changes now

ayushnawal committed:
return on pygame.QUIT

chimosky commented:
Tested, activity freezes when `Escape` is pressed also occurs at master.
Display doesn't fit screen size - 1366x694 -.

quozl commented:
> Tested, activity freezes when `Escape` is pressed also occurs at master.

Yeah, looking at `game/__init__.py` the coded response to escape is not correct.

ayushnawal commented:
@quozl @chimosky I looked closely at the code, I am very much confused because the code for `escape` and `alt + F4` is similar and for alt + F4 it is working as expected but not for escape. It is coming out of the game loop after pressing but there is some issue with the cleanup.

However adding sys.exit and pygame.quit in response to escape fixed the freezing error.

Does escape have some bindings in sugar that I am unaware of and need to be taken care of?

@srevinsaju do you have any idea?

quozl commented:
Thanks for looking into this.

So this pull request began with your wish to update from Sugargame v1.2 to Sugargame v1.3, and this gave us a chance to review the use of Sugargame and how the activity responds.  The activity was a team project at Rochester Institute of Technology (RIT) and was published in May 2016 as [experimental](http://activities.sugarlabs.org/en-US/sugar/addon/4790).  After further development it was stalled in 2017 by a black screen issue; which was caused by initialising Pygame at the wrong time; which could have been a Sugargame v1.1 hold-over.  My port to Sugargame v1.2 was in 2018.

The issues we've seen in this pull request look similar to issues seen in all activities that are ported from Sugargame v1.1 and the older versions of GTK.  Calling sys.exit or pygame.quit is an unacceptable fix because it doesn't allow Sugar Toolkit to clean up, and we've talked about that at length in previous pull requests.  Make sure you keep all notifications somewhere so you can search your mail, or search GitHub for your closed pull requests.

To answer your confusion, in the code as it stands in your branch
* the escape key down event sets the flag quit_attempt,
* the F4 key down event while either alt key is pressed sets the flag quit_attempt,
* on the next clock tick (an unnecessary delay), the call to poll_cb will detect quit_attempt and call Gtk.main_quit,
* the GTK main loop will exit, but without closing the activity.

The Sugargame test activity gives a clear example of how to handle this;
* there is no alt+f4 or escape handler,
* the stop button is visible,
* ctrl+q activates the stop button,
* the stop button calls Activity.close (see Sugar Toolkit class StopButton).

The other important thing that @chimosky said was that the activity does not fit a display.

ayushnawal committed:
remove other quit shortcuts

ayushnawal commented:
handled it after reviewing test activity.

tested. quitting on stop or `ctrl-q` now. 

@quozl @chimosky, please review

chimosky commented:
Tested 9838573, works as expected. Display issue still persists.

quozl commented:
Reviewed 9838573 alone.  I don't know why you removed the other quit shortcuts.  Why did you do that?  I saw you changed the stop button callback to use quit_attempt instead of running, which was good, but you didn't mention that in the commit message.

After your change, from looking at the code in both files;
* in response to ctrl+q or the stop button being pressed, _poll_cb will call Gtk,main_quit, but the event loop will continue to execute.

But the test activity handles ctrl+q or the stop button by leaving the event loop.

ayushnawal committed:
Fix quit method of activity

ayushnawal commented:
> Display issue still persists.

I tried it again. unsuccessful. @chimosky or @quozl  can you please guide me here?

> I don't know why you removed the other quit shortcuts

There is no alt+f4 or escape handler in test activity so I removed it from here also, to try to keep the method of quitting any activity consistent.

> in response to ctrl+q or the stop button being pressed, _poll_cb will call Gtk,main_quit, but the event loop will continue to execute.

Fixed it now, _poll_cb was not required actually, now the activity quits just like our test activity by leaving the event loop

quozl commented:
> > Display issue still persists.
> 
> I tried it again. unsuccessful. @chimosky or @quozl can you please guide me here?

I've no specific guidance.  The activity was written for a laptop with 1200x900 display resolution, which you can see also in the PyCut.py test harness that was used for running the activity outside Sugar, and in how the width and height were hard coded in the PyCutGame initialiser.  You can also see how two of the imagery assets were prepared for the exact resolution, instead of being scalable vector graphics (SVG) with rendering delayed until resolution is known.

General guidance would be to apply software design and coding techniques, and go _back_ and repeat the steps that were taken, but this time;

* [ ] convert the bitmap graphics to scalable vector graphics, (we have graphic artists who can get involved, @perriefidelis can advise, or ask on sugar-devel@),
* [ ] detect the resolution of the display instead of hard coding it,
* [ ] render the graphics at a scale in proportion to the original design resolution,
* [ ] detect changes to the resolution and render the graphics again at a new scale.

> > I don't know why you removed the other quit shortcuts
> 
> There is no alt+f4 or escape handler in test activity so I removed it from here also, to try to keep the method of quitting any activity consistent.

I don't think that's a good enough reason.  There will be people who have used the activity and learned the extra shortcuts, and forcing them to change just because you want it to be like something else doesn't seem right.  ctrl+q is consistent.

> > in response to ctrl+q or the stop button being pressed, _poll_cb will call Gtk,main_quit, but the event loop will continue to execute.
> 
> Fixed it now, _poll_cb was not required actually, now the activity quits just like our test activity by leaving the event loop

That's a pity.  The new mechanism of _poll_cb that I wrote helped by keeping the Pygame program from having to import GTK.  It is something I'm considering for the next release of Sugargame, along with a way to avoid use of that instance variable self.game.running.
